clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -0,0 +1,449 @@
+        return value == null ? null : value.duplicate();
+    }
+
+    @Override
+    public Header[] headers() {
+        return headers;
+    }
+
+    public static long writeTo(DataOutputStream out,
+                               boolean isControlRecord,
+                               int offsetDelta,
+                               long timestampDelta,
+                               ByteBuffer key,
+                               ByteBuffer value,
+                               Header[] headers) throws IOException {
+        int sizeInBytes = sizeOfBodyInBytes(offsetDelta, timestampDelta, key, value, headers);
+        ByteUtils.writeVarint(sizeInBytes, out);
+
+        byte attributes = computeAttributes(isControlRecord);
+        out.write(attributes);
+
+        ByteUtils.writeVarlong(timestampDelta, out);
+        ByteUtils.writeVarint(offsetDelta, out);
+
+        if (key == null) {
+            ByteUtils.writeVarint(-1, out);
+        } else {
+            int keySize = key.remaining();
+            ByteUtils.writeVarint(keySize, out);
+            out.write(key.array(), key.arrayOffset(), keySize);
+        }
+
+        if (value == null) {
+            ByteUtils.writeVarint(-1, out);
+        } else {
+            int valueSize = value.remaining();
+            ByteUtils.writeVarint(valueSize, out);
+            out.write(value.array(), value.arrayOffset(), valueSize);
+        }
+
+        if (headers == null)
+            throw new IllegalArgumentException("Headers cannot be null");
+
+        ByteUtils.writeVarint(headers.length, out);
+
+        for (Header header : headers) {
+            String headerKey = header.key();
+            byte[] utf8Bytes = Utils.utf8(headerKey);
+            ByteUtils.writeVarint(utf8Bytes.length, out);
+            out.write(utf8Bytes);
+
+            ByteBuffer headerValue = header.value();
+            if (headerValue == null) {
+                ByteUtils.writeVarint(-1, out);
+            } else {
+                int headerValueSize = headerValue.remaining();
+                ByteUtils.writeVarint(headerValueSize, out);
+                out.write(headerValue.array(), headerValue.arrayOffset(), headerValueSize);
+            }
+        }
+
+        return computeChecksum(timestampDelta, key, value);
+    }
+
+    public static long writeTo(ByteBuffer out,
+                               boolean isControlRecord,
+                               int offsetDelta,
+                               long timestampDelta,
+                               ByteBuffer key,
+                               ByteBuffer value,
+                               Header[] headers) {
+        try {
+            return writeTo(new DataOutputStream(new ByteBufferOutputStream(out)), isControlRecord, offsetDelta,
+                    timestampDelta, key, value, headers);
+        } catch (IOException e) {
+            // cannot actually be raised by ByteBufferOutputStream
+            throw new RuntimeException(e);
```

Comment:
Could we include an error message to indicate that this is unexpected? Or could we throw an IllegalStateException instead?